docs/src/main/asciidoc/cdi-integration.adoc

```diff
@@ -419,6 +419,71 @@ public class TestRecorder {
 ----
 <1> Pass a contextual reference of `Bar` to the constructor of `Foo`.
 
+=== Inactive Synthetic Beans
+
+In the case when one needs to register multiple synthetic beans at build time but only wants a subset of them active at runtime, it is useful to be able to mark a synthetic bean as _inactive_.
+This is done by configuring a "check active" procedure, which should be a `Supplier<ActiveResult>` obtained from a recorder:
+
+.Inactive Synthetic Bean - Build Step Example
+[source,java]
+----
+@BuildStep
+@Record(RUNTIME_INIT)
+SyntheticBeanBuildItem syntheticBean(TestRecorder recorder) {
+    return SyntheticBeanBuildItem.configure(Foo.class)
+            .scope(Singleton.class)
+            .startup() // <1>
+            .checkActive(recorder.isFooActive()) // <2>
+            .createWith(recorder.createFoo())
+            .done();
+}
+----
+<1> A bean that might be inactive is typically initialized eagerly, to make sure that an error is thrown at application startup.
+    If the bean is in fact inactive, but is not injected into an always-active bean, eager initialization is skipped and no error is thrown.
+<2> Configures the "check active" procedure.
+
+.Inactive Synthetic Bean - Recorder Example
+[source,java]
+----
+@Recorder
+public class TestRecorder {
+    public Supplier<ActiveResult> isFooActive() {
+        return () -> {
+            if (... should not be active ...) { // <1>
+                return ActiveResult.inactive("explanation"); // <2>
+            }
+            return ActiveResult.active();
+        };
+    }
+
+    public Function<SyntheticCreationalContext<Foo>, Foo> createFoo() {
+        return (context) -> {
+            return new Foo();
+        };
+    }
+}
+----
+<1> The condition when the synthetic bean should be inactive.
+<2> Proper explanation of why the bean is inactive.
+    Another inactive `ActiveResult` may also be provided as a cause, if this bean's inactivity stems from another bean's inactivity.
+
+If an inactive bean is injected somewhere, or is dynamically looked up, an `InactiveBeanException` is thrown.
+The error message contains the reason (from the `ActiveResult`), the cause chain (also from the `ActiveResult`), and possibly also a list of all injection points that resolve to this bean.
+
+This means that a possibly inactive bean should only be injected using `Instance<\...>`.
```

Comment:
We could do that during build, yes. If someone directly injects a bean that is possibly-inactive, that is probably an error.

Currently, we don't -- we only do that at runtime.

Reply:
Personally, I think we should enforce it.
ATM we are saying that people should do that anyway and we assume they need to check if the bean is active - both parts can be part of an error msg user gets if they try to perform a direct injection of such a bean.

Reply:
Yeah, agree. I'll try to add a build-time validation.

Reply:
Wait, datasources are always "possibly inactive". Are we really talking about forbidding users from injecting datasources without wrapping them in `Instance<>`? That seems... extreme?

Please keep in mind that while beans such as datasources are "possibly inactive", in most applications it will be very unlikely:

1. While datasources expose a ".active" runtime configuration property, the vast majority of users won't use it or even know it's there.
2. The only remaining way for a datasource to be inactive is the result of a (runtime) configuration mistake, i.e. forgetting to set the datasource URL.

So I don't think it's a good idea to fail at build time if a "possibly inactive" bean is injected without being wrapped in `Instance<>`.

Reply:
Ouch. Yeah, you're right, that does seem excessive.

I'll try to highlight this in the documentation. If you want to handle the inactive case gracefully, you should always inject using `Instance<>`. If you want the inactive case to blow up, just use regular injection.

Reply:
Fair enough.
We should then avoid recommending `Instance` as a universal approach.

Reply:
Fixed the documentation.

Reply:
I believe this is done now.